[audio] TrueHD rework

thexai opened the pull request:
## Description
TrueHD rework

Closes https://github.com/xbmc/xbmc/issues/20391 (before was fixed only IEC, this PR fixes both RAW & IEC)

## Motivation and context
My understanding is that with TrueHD the water level constantly fluctuates around the `MAX_WATER_LEVEL` a little above and a little below but that's okay. The only problem is that AE was not designed with this in mind.

Changing constant value of `MAX_WATER_LEVEL` to other number does not fix the issue because still fluctuates (in other target level).

Note that this issue is **very old**: already **was happening in v18.9** and older and is unrelated to recent TrueHD rework to support high bitrate TrueHD (new MAT packer) and **is unrelated to the fact of use 12 or 24 audio units**.

**Confirmed things**

- This is the root cause of PAPlayer progress indicator jumps.
- And more important: this is also one of the causes of video stuttering in Android ("smoothing errors" also contributes here).  Basically Android stuttering issues can be avoided if the duration is falsified with the "ratio" trick (that it has been seen that it is not technically correct because the TrueHD MAT frames are 20ms on Sink) OR "patching" water level. This maybe is another workaround but is more near of true root cause.
- PAPlayer jumps in progress issue exists from before TrueHD MAT packer first rework in v19.3. See: https://github.com/xbmc/xbmc/issues/20391 Then is unrelated to use 12 or 24 audio units and even is unrelated to use IEC or RAW.

> Kodi version:
> 19.0, 19.1, 19.2
> 18.4, 18.5 etc


**Proposed Solution**

Root cause is that a/v sync only is calculated when current water level is < `MAX_WATER_LEVEL`:

https://github.com/xbmc/xbmc/blob/fa592d344eb0b5e2bc8864cb304173e696e36a79/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L1954-L1957

Then in TrueHD a/v sync is NOT calculated all the time and this causes issues mentioned above.

Proposed solution is calculate a/v sync error all the time in TrueHD only (regardless of water level).

Note that in any case water level is always less than `MAX_CACHE_LEVEL` then no side effects are expected.


## How has this been tested?
Runtime tested Windows 11 22H2 (Intel NUC8i3BEK + Denon X1600H)
Runtime tested Android (Shield FW 9.1.1 + Denon X1600H)

None issues found: absolute no TrueHD audio dropouts. Video super smooth. TrueHD a/v sync perfect. Debug info OSD better than ever a/v sync fluctuates a little around zero, perfectly stable. 

See: https://user-images.githubusercontent.com/58434170/217493470-d26ee717-dde6-4bda-9353-a3cabad78eac.mp4

For me it is the best version for Shield that has ever existed.


## What is the effect on users?
The differences are subtle as both master and Nexus already include TrueHD workarounds. Now has been perfected, everything is a little better. The only major difference is that this is the first and only Kodi version where the PAPlayer progress bar works well with TrueHD AudioTrack RAW.


## Screenshots (if appropriate):

**Before - Current master branch - AudioTrack RAW**
https://user-images.githubusercontent.com/58434170/217491726-1e10aebb-8ac6-4b0a-9e7f-d9c11abb5188.mp4

**After - This PR - AudioTrack RAW**
https://user-images.githubusercontent.com/58434170/217492052-7fa8435e-6fc0-42fc-bb13-0543c0b76037.mp4


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai committed:
ActiveAE: use float for "water level" constants and remove unnecessary static_cast

thexai committed:
AEStreamInfo: drop TrueHD GetDuration() special case

thexai committed:
AESinkAUDIOTRACK: remove TrueHD IEC "ratio" workaround

thexai committed:
ActiveAE: take into account TrueHD MAT packer latency

CastagnaIT commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -29,9 +29,12 @@ using namespace ActiveAE;
-#define MAX_BUFFER_TIME 0.1   // max time of a buffer in seconds
+namespace
+{
+constexpr auto MAX_CACHE_LEVEL = 0.4f; // total cache time of stream in seconds;
+constexpr auto MAX_WATER_LEVEL = 0.2f; // buffered time after stream stages in seconds;
+constexpr auto MAX_BUFFER_TIME = 0.1; // max time of a buffer in seconds;
```
can we avoid abuse of `auto` expression?
when there are many variables is better readable when types are explicit set

fritsch replied:
I would also use explicit float here - especially in combination with the forced type at the end.
Another idea to care for rounding errors is using double if that is needed.

thexai replied:
> Another idea to care for rounding errors is using double if that is needed.

Use double for MAX_WATER_LEVEL constant and level calculation does not change things. Exact same numbers obtained.

fritsch replied:
Jep - I also think that a little epsilon helps, especially if the delta is below one packet - it is clearly a rounding error somewhere in the middle.

fritsch commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -152,6 +152,13 @@ void CEngineStats::GetDelay(AEDelayStatus& status, CActiveAEStream *stream)
+      m_sinkFormat.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD)
+  {
+    // take into account MAT packer latency
+    status.delay += 12.0 * m_sinkFormat.m_streamInfo.GetDuration() / 1000;
```
This here results in 240 ms. Isn't it rather (12.0 / 24.0), e.g. 1/2 of 20 ms?

thexai replied:
You're right, I think I've somehow related it to the 250 ms commented on the forum, but it has to be something else.

12 + 12 + 12 -----> MAT packer ----> output 24

latency is 12 units which is 10 ms (half MAT package)


Maybe exact 200 ms is the "WATER_LEVEL" duration of buffer (when is filled with 10 MAT packages or 10 x 24 audio units)

fritsch replied:
Okay, that maybe means we are on average 12 units off, so around:
```
    status.delay += 0.5 * m_sinkFormat.m_streamInfo.GetDuration() / 1000;
```
But only for IEC, cause for RAW we package 24 at once.

fritsch replied:
But actually also only once. Cause after that always 24 units can be output, just with 12 units (10 ms) delay.

thexai replied:
RAW packer is transparent with no latency

24 units ----->  RAW packer ----> output 24 units in the same function call (no latency)

fritsch replied:
Jep - exactly. That's why it's only relevant for the very first package and only for IEC. Code is correct as you have it right now!

thexai replied:
Now is:
`status.delay += m_sinkFormat.m_streamInfo.GetDuration() / 1000 / 2;`

should be equivalent although i can change

thexai replied:
One important thing here is that I cannot reproduce any delay in my setup changing delay here (even 250 ms) but maybe for users that have this issue, can be compensated here (only for TrueHD IEC) and works fine for all. 

As seems not affect to users that NOT have a/v sync issues.

fritsch commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -217,19 +220,28 @@ float CEngineStats::GetCacheTotal()
+               1000;
+
+    if (wl >= MAX_WATER_LEVEL)
+      wl = MAX_WATER_LEVEL - 0.00001f;
```
Don't you want to return wl - 0.000001f to compensate for rounding errors. In error case there might be more far more buffered samples in there and AE would not realize.

fritsch replied:
```
if (wl > MAX_WATER_LEVEL && wl < MAX_WATER_LEVEL + EPSILON)
  wl = MAX_WATER_LEVEL;
```

thexai commented:
Something not works in AudioTrack part:

Delay-Offset: 9.607679205057056e+16 ms

```
2023-02-05 17:47:46.007 T:8487     info <general>: Framecounter: 217076 Time: 275891666977 Current-Time: 276155857290
2023-02-05 17:47:46.007 T:8487     info <general>: Delay - Timestamp: 305.08854166666777 (ms) delta: 264.308021 (ms) playtime: 1394.9114583333333 (ms) Duration: 1700.0000000000011 ms
2023-02-05 17:47:46.007 T:8487     info <general>: Head-Position 264172 Timestamp Position 267823 Delay-Offset: 9.607679205057056e+16 ms <------------------
2023-02-05 17:47:46.007 T:8487     info <general>: HW-Delay (1): -19.01562499999998 ms
2023-02-05 17:47:46.007 T:8487     info <general>: Combined Delay: 305.08854166666777 ms
2023-02-05 17:47:46.007 T:8487     info <general>: Delay Current: 299.231771 ms
2023-02-05 17:47:46.127 T:8487     info <general>: Framecounter: 217076 Time: 275891666977 Current-Time: 276276599581
2023-02-05 17:47:46.128 T:8487     info <general>: Delay - Timestamp: 204.312500000001 (ms) delta: 385.086146 (ms) playtime: 1515.6875 (ms) Duration: 1720.0000000000011 ms
2023-02-05 17:47:46.128 T:8487     info <general>: Head-Position 290808 Timestamp Position 291012 Delay-Offset: **9.607679205057058e+16 ms**
2023-02-05 17:47:46.128 T:8487     info <general>: HW-Delay (1): -1.0624999999999662 ms
2023-02-05 17:47:46.128 T:8487     info <general>: Combined Delay: 204.312500000001 ms
2023-02-05 17:47:46.128 T:8487     info <general>: Delay Current: 253.237847 ms
2023-02-05 17:47:46.129 T:8487     info <general>: Framecounter: 217076 Time: 275891666977 Current-Time: 276277867394
2023-02-05 17:47:46.129 T:8487     info <general>: Delay - Timestamp: 223.08333333333442 (ms) delta: 386.314375 (ms) playtime: 1516.9166666666667 (ms) Duration: 1740.0000000000011 ms
2023-02-05 17:47:46.129 T:8487     info <general>: Head-Position 290808 Timestamp Position 291248 Delay-Offset: 9.607679205057058e+16 ms
2023-02-05 17:47:46.129 T:8487     info <general>: HW-Delay (1): -2.291666666666581 ms
2023-02-05 17:47:46.129 T:8487     info <general>: Combined Delay: 223.08333333333442 ms
2023-02-05 17:47:46.129 T:8487     info <general>: Delay Current: 230.493924 ms
2023-02-05 17:47:46.130 T:8487     info <general>: Framecounter: 217076 Time: 275891666977 Current-Time: 276279089685
2023-02-05 17:47:46.130 T:8487     info <general>: Delay - Timestamp: 241.88020833333445 (ms) delta: 387.51974 (ms) playtime: 1518.1197916666667 (ms) Duration: 1760.0000000000011 ms
2023-02-05 17:47:46.130 T:8487     info <general>: Head-Position 290808 Timestamp Position 291479 Delay-Offset: 9.607679205057058e+16 ms
2023-02-05 17:47:46.130 T:8487     info <general>: HW-Delay (1): -3.4947916666665524 ms
2023-02-05 17:47:46.130 T:8487     info <general>: Combined Delay: 241.88020833333445 ms
2023-02-05 17:47:46.130 T:8487     info <general>: Delay Current: 229.353299 ms
2023-02-05 17:47:46.131 T:8487     info <general>: Framecounter: 217076 Time: 275891666977 Current-Time: 276280208435
2023-02-05 17:47:46.131 T:8487     info <general>: Delay - Timestamp: 260.76562500000125 (ms) delta: 388.631615 (ms) playtime: 1519.234375 (ms) Duration: 1780.0000000000011 ms
2023-02-05 17:47:46.131 T:8487     info <general>: Head-Position 290808 Timestamp Position 291693 Delay-Offset: 9.607679205057058e+16 ms
2023-02-05 17:47:46.131 T:8487     info <general>: HW-Delay (1): -4.609374999999805 ms
2023-02-05 17:47:46.131 T:8487     info <general>: Combined Delay: 260.76562500000125 ms
```

I think I'm going to close the PR because the tests are not conclusive and it doesn't work better than what's already out there.

fritsch commented:
That might be okay:
Head-Position 264172 Timestamp Position 267823

The HeadPosition of the old Audiotrack API did not move yet, while the TimeStamp API already moved. The issue is that the "Difference" is computed in uint64_t - and not properly casted to an int64_t - that's basically everything.

```
1000.0 * (static_cast<int64_t)(m_headPos - m_timestampPos)) / m_sink_sampleRate);
```

fritsch committed:
AESinkAudioTrack: Fix Print for TimeStampDelay

When the original HeadPosition did not yet move the TimeStamp which is
interpolated will already have moved. The print worked on unsigned and
did not take this case into account. Fix it by a cast.

fritsch commented:
Sent you a pullrequest: https://github.com/thexai/xbmc/pull/2

thexai commented:
Seems this happens also in current master, nothing related with PR

Others files works fine:

```
2023-02-05 18:23:36.232 T:12967    info <general>: Framecounter: 493483 Time: 2426186844230 Current-Time: 2426381390792
2023-02-05 18:23:36.232 T:12967    info <general>: Delay - Timestamp: 190.7482993196652 (ms) delta: 194.678802 (ms) playtime: 11384.761904761906 (ms) Duration: 11575.510204081571 ms
2023-02-05 18:23:36.232 T:12967    info <general>: Head-Position 505990 Timestamp Position 502068 Delay-Offset: 88.93424036281179 ms
2023-02-05 18:23:36.232 T:12967    info <general>: HW-Delay (1): 88.93424036281061 ms
2023-02-05 18:23:36.233 T:12967    info <general>: Combined Delay: 190.7482993196652 ms
2023-02-05 18:23:36.233 T:12967    info <general>: Delay Current: 180.574452 ms
2023-02-05 18:23:36.299 T:12967    info <general>: Framecounter: 493483 Time: 2426186844230 Current-Time: 2426448525271
2023-02-05 18:23:36.299 T:12967    info <general>: Delay - Timestamp: 155.75963718814558 (ms) delta: 261.815208 (ms) playtime: 11451.904761904761 (ms) Duration: 11607.664399092908 ms
2023-02-05 18:23:36.300 T:12967    info <general>: Head-Position 508870 Timestamp Position 505029 Delay-Offset: 87.09750566893425 ms
2023-02-05 18:23:36.300 T:12967    info <general>: HW-Delay (1): 87.09750566893426 ms
2023-02-05 18:23:36.300 T:12967    info <general>: Combined Delay: 155.75963718814558 ms
2023-02-05 18:23:36.300 T:12967    info <general>: Delay Current: 170.699169 ms
2023-02-05 18:23:36.316 T:12967    info <general>: Framecounter: 493483 Time: 2426186844230 Current-Time: 2426465266782
2023-02-05 18:23:36.316 T:12967    info <general>: Delay - Timestamp: 171.20181405889312 (ms) delta: 278.534271 (ms) playtime: 11468.616780045351 (ms) Duration: 11639.818594104245 ms
2023-02-05 18:23:36.316 T:12967    info <general>: Head-Position 509590 Timestamp Position 505766 Delay-Offset: 86.71201814058956 ms
2023-02-05 18:23:36.316 T:12967    info <general>: HW-Delay (1): 86.71201814058982 ms
2023-02-05 18:23:36.316 T:12967    info <general>: Combined Delay: 171.20181405889312 ms
2023-02-05 18:23:36.316 T:12967    info <general>: Delay Current: 169.312169 ms
```

fritsch commented:
It's entirely fine. The Delay copes with it. Only the print is wrong.

fritsch commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -217,19 +220,31 @@ float CEngineStats::GetCacheTotal()
+    if (m_sinkFormat.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD)
+    {
+      if (wl >= MAX_WATER_LEVEL)
+        wl = MAX_WATER_LEVEL - 0.00001f;
```
I would add a sanity check as well to see if TrueHD actually overruns MAX_WATER_LEVEL by far more milliseconds.

fritsch replied:
The more I check the code, the more I get the feeling that the waterlevel is not relevant for TrueHD. For me it seems that Paplayer is the only guy doing something special here with the levels. See also forum feedback without that change - I will look more in detail over the next few days. Cutting an "arbitrary" high value down to WATER_LEVEL is not good - it's another workaround for something we don't have yet fully understood.

thexai replied:
Already explained but anyway:

Shield has a very specific micro-stutter type _only_ with TrueHD.  (some people has also other types of stuttering due high CPU but are unrelated to this).

Micro-stutter of "TrueHD only" can be fixed with "ratios" in AudioTrack and half `GetDutation()`

Now something NEW has been discovered:

These micro-stutter of "TrueHD only" ALSO are gone reverting ratio and GetDuration() BUT changing water level behavior.

Seems you not like water level and propose revert also ratio and GetDuration().... 

Then inevitably you get the TrueHD micro-stutter again and Shield is not usable.

thexai replied:
>  See also forum feedback without that change

Because in forum people are using advancedsettings with value 90 -100 that only should be used to fix the others type of video stuttering (due high CPU usage). Also this setting introduces a/v sync issues of same magnitude that value of setting, so is not usable for everyone.

thexai replied:
from forum:
> That build only changes one thing to yesterday: It uses 24 audio units for TrueHD again

This is not usable because re-introduce random audio dropouts due MAT packer latency:

24 + 24 ----->  MAT packer ------> 24

there are 24 units of latency 20ms vs 10ms using  12 + 12

fritsch replied:
I agree for the last comment: 12 Units it needs to be. Forum has shown that.
For the first in correlation with the discon - I don't. This is needed for them no matter if ratio is applied or not. It only caused in my POV from the large SINK error that is caused by nvidia's AudioTrack implementation: 1,2,1,80 ms blocking time.

thexai replied:
> This is needed for them no matter if ratio is applied or not.

Right but this is separated issue (not related with TrueHD) and only some users are affected by this.

In my setup I have micro-stutter free playback with default 10ms ONLY if water level modification is applied.

thexai replied:
**option A:** water level workaround and ratio / half duration reversion (this PR)

**option B:** close PR and remains ratio / half duration as current master & Nexus

**option C:** no water level workaround and ratio / half duration reversion. => **NOT POSSIBLE**

**option D:** use 24 audio units in MAT packer => **NOT POSSIBLE**

**option E:** fix true root cause of PAPlayer progress indicator and "a/v sync oscillation" and debug info OSD oscillation. ??

Note that same thing happens in PAPlayer happens also in VideoPlayer, only difference is that is not visible because progress indicator is based on video time stamps (not audio). Then is "not visible" but same issue exists.

fritsch replied:
I want to look into the DVD Passthrough Codec this evening. Here the PTS timestamps are "computed" ... hopefully they don't role forth and back ...

If you check where WaterLevel is basically used in AE, the you see that passthrough is the "else" case ... where it should not play a role. In PaPlayer the audio does not stutter, but we see that the playbar is jumpy - as PaPlayer updates the values by itself (setting his status) most likely we can find out, why the value flactuates.

To see if E is really an option.

For the ratio workaround: ALSA and Windows need adjusting ... as well, this is currently not correct then, only AudioTrack works around this. As show on the forum: Impact of 100 ms further delay, cause AE does not know anymore, that Sink has 100 ms audio more than it reports ... this is why I don't like the ratio workaround.

fritsch replied:
For PAPlayer:
```
      si->m_framesSent += si->m_audioFormat.m_streamInfo.GetDuration() / 1000 *
                          si->m_audioFormat.m_streamInfo.m_sampleRate;
```
```
free_space = (double)si->m_stream->GetSpace() *
                     si->m_audioFormat.m_streamInfo.GetDuration() / 1000;
```

The first one is always: 0.02 * 192 khz - so both assume again that every frame contains 20 ms of data.
The second one is always: GetSpace x 0.02 - but there is more: the first method assume a direct relation ship between "payload time" and samplerate, which is only the cause if the payload of a frame matches the duration of the data. The second method instead only again is intersted in the number of free sample by duration.

While GetSpace is from the Stream itself:
```
unsigned int CActiveAEStream::GetSpace()
{
  std::unique_lock<CCriticalSection> lock(m_streamLock);
  if (m_format.m_dataFormat == AE_FMT_RAW)
    return m_streamFreeBuffers;
  else
    return m_streamFreeBuffers * m_streamSpace;
}

The m_streamFreeBuffers are incremented whenever possible. AE is blocking below to fillup the stream again. until the cache has MAX_CACHE_LEVEL is reached. So being larger than CACHE_LEVEL is a good thing.
```
    // provide buffers to stream
    float time = m_stats.GetCacheTime((*it));
    CSampleBuffer *buffer;
    if (!(*it)->m_drain)
    {
      float buftime = (float)(*it)->m_inputBuffers->m_format.m_frames / (*it)->m_inputBuffers->m_format.m_sampleRate;
      if ((*it)->m_inputBuffers->m_format.m_dataFormat == AE_FMT_RAW)
        buftime = (*it)->m_inputBuffers->m_format.m_streamInfo.GetDuration() / 1000;
      while ((time < static_cast<float>(MAX_CACHE_LEVEL) || (*it)->m_streamIsBuffering) &&
             !(*it)->m_inputBuffers->m_freeSamples.empty())
      {
        buffer = (*it)->m_inputBuffers->GetFreeBuffer();
        (*it)->m_processingSamples.push_back(buffer);
        (*it)->m_streamPort->SendInMessage(CActiveAEDataProtocol::STREAMBUFFER, &buffer, sizeof(CSampleBuffer*));
        (*it)->IncFreeBuffers();
        time += buftime;
      }
    }
```
And with this very same assumption paplayer clustters this in its time info, here:
```
  double time = ((double)m_currentStream->m_framesSent / (double)m_currentStream->m_audioFormat.m_sampleRate);
  if (m_currentStream->m_stream)
    time -= m_currentStream->m_stream->GetDelay();
  time = time * 1000.0;

  m_playerGUIData.m_time = (int64_t)time; //update for GUI
 ```

fritsch replied:
And bye -> lunchbreak is over :-)

thexai replied:
Other important point:

PAPlayer jumps in progress issue exists from before TrueHD MAT packer first rework in v19.3. See: https://github.com/xbmc/xbmc/issues/20391 Then is unrelated to use 12 or 24 audio units and even is unrelated to use IEC or RAW.

> Kodi version:
> 19.0, 19.1, 19.2
> 18.4, 18.5 etc

This means that what we are trying to patch here is a bug in AE core that has always been latent.

Tomorrow try to split water level commit in two:

first one to fix float "rounding" issues, that is

`if (float A == float B) ` is incorrect by design

`if (abs(A - B) < epsilon)`    <------ should be

in a second commit specific behavior changes for TrueHD only (this can be reverted when a better solution is found)

fritsch replied:
I think we can ommit the second commit for now and just let it jump - and then properly fixing paplayer. From my investigation the problem is with framesent using samplerate for conversion and freespace the duration ... so one is tied to the size of the package that other not. This can't work.

AE does not check == but use while (time < MAXLEVEL) which is perfectly fine, as this will end when it is <= MAXLEVEL.

thexai replied:
Sorry, but you are not understanding the problem. PAPlayer doesn't matter to me. What I want to fix are the problems in VideoPlayer with TrueHD

fritsch replied:
Have a look here: 09b0276868c413a168c9cbc640c878d737fe1836

thexai replied:
Same BUG causes:

- PAPlayer jumps.
- Debug info OSD a/v sync fluctuates much with TrueHD  (playing video).
- Android (Shield) micro-stutter playing video + TrueHD.

ALL THIS is due water level issue

fritsch replied:
Please see the second part of the commit I just linked. Here the TrueHD special case needs to go in again and only half the duration needs to be added when the 12 frame version is running.

fritsch replied:
The original code added the timing by 1/24 ... but then it was changed to only once per 61440 add one frame. Now we come twice - therefore adding half of it - should fix the timing again.

fritsch replied:
```
      bool rawPktComplete = false;
      {
        std::unique_lock<CCriticalSection> lock(m_statsLock);
        if (m_format.m_dataFormat != AE_FMT_RAW)
        {
          m_currentBuffer->pkt->nb_samples += minFrames;
          m_bufferedTime +=
              static_cast<float>(minFrames) / m_currentBuffer->pkt->config.sample_rate;
        }
        else
        {
          if (m_format.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD && isIEC)
          {
            m_currentBuffer->pkt->nb_samples += 30720;
            m_bufferedTime += m_format.m_streamInfo.GetDuration() / 1000 / 2;
            if (m_currentBuffer->pkt->nb_samples / 2560 == 24)
              rawPktComplete = true;
          }
          else
          {
            m_bufferedTime += static_cast<float>(m_format.m_streamInfo.GetDuration()) / 1000;
            m_currentBuffer->pkt->nb_samples += minFrames;
            rawPktComplete = true;
          }
        }
      }
 ```
 something like this, that way the m_bufferedTime will only increase half and the rest of the condition can stay as is ...

thexai replied:
This looks promising 👍   I'll try it when I have time

Thank you!

fritsch replied:
Nothing to thank - we have the same goal :-)

thexai replied:
Unfortunately it doesn't work, some reasons I understand others I don't:

`rawPktComplete = true;` needs in all half packets because in

https://github.com/xbmc/xbmc/blob/fa592d344eb0b5e2bc8864cb304173e696e36a79/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp#L951

needs to processed every half packet (12 units). Then `ActiveAESink` half packet is considered as `rawPktComplete`

`m_format.m_streamInfo.GetDuration() / 1000 / 2;`  seems very logic and correct but causes video player hangs (not starts playback any video frame) / black screen.

PAPlayer works but not resolves main issues: progress indicator stills jumps and duration is wrong (speed x2). Audio is OK.

thexai replied:
See comment in a new commit https://github.com/xbmc/xbmc/pull/22705/commits/dbbb535d272f20d652b37cc3bb5d55fc4890c1c4 to understand why jumps progress indicator:

jumps is because a/v sync in NOT calculated all the time because condition 

`m_stats.GetWaterLevel() < MAX_WATER_LEVEL  `

is not true all the time.

thexai replied:
My understanding is that with TrueHD the water level constantly fluctuates around the `MAX_WATER_LEVEL`: a little above and a little below but that's okay. The only problem is that AE was not designed with this in mind.

fritsch replied:
> Unfortunately it doesn't work, some reasons I understand others I don't:
> 
> `rawPktComplete = true;` needs in all half packets because in
> 
> https://github.com/xbmc/xbmc/blob/fa592d344eb0b5e2bc8864cb304173e696e36a79/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp#L951
> 
> needs to processed every half packet (12 units). Then `ActiveAESink` half packet is considered as `rawPktComplete`
> 
> I mean that `ActiveAESink` only is called half of times and audio is corrupt.
> 
> `m_format.m_streamInfo.GetDuration() / 1000 / 2;` seems very logic and correct but causes VideoPlayer hangs (not starts playback any video frame) / black screen.
> 
> PAPlayer works but not resolves main issues: progress indicator stills jumps and duration is wrong (speed x2). Audio is OK.

Okay - that actually makes sense. We pull out 12 units all the time, but our IEC packer, the final guy in the chain at the end only outputs something after 24 packets were added. So therefore faking "Complete" to VideoPlayer all the time could be done, but then only adding half of the time?

Like:
```
      {
        std::unique_lock<CCriticalSection> lock(m_statsLock);
        if (m_format.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD && isIEC)
        {
          m_currentBuffer->pkt->nb_samples += minFrames;
          m_bufferedTime += m_format.m_streamInfo.GetDuration() / 1000 / 2;
          rawPktComplete = true;
        }
        else
        {
        .... // current code
        }
      }
 ```

That should achieve that rawPktComplete as always true to enter next stage, but not adding a wrong delay?

thexai replied:
As commented 

`m_bufferedTime += m_format.m_streamInfo.GetDuration() / 1000 / 2;`

causes VideoPlayer hangs / black screen

fritsch replied:
No hurry. Then on the other side of things, when running through "the stages" someone still has the wrong information ... will also check again on the weekend.

thexai replied:
Code to clamp water level to MAX_WATER_LEVEL has been removed, then this issue is closed.

@fritsch, please review as new PR.  

Open your mind and do not bring preconceived ideas 😊

fritsch commented:
Changes look good. Would be nice if you could pick https://github.com/thexai/xbmc/pull/2/commits/aa1b6d62d1bae1018e6308cb564a624900c512e1 - that will directly stop the confusing print in AudioTrack as you touch it anyways.

thexai committed:
ActiveAE: ignore "water level" in TrueHD, calculate a/v sync error all the time

As water level fluctuates more with TrueHD there are instants when a/v sync
error is not calculated, i.e:

GetWaterLevel() > MAX_WATER_LEVEL

With this change a/v sync is calculated all the time.

thexai committed:
PAPlayer: reintroduces special case for TrueHD IEC

fritsch commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -1911,7 +1913,7 @@ bool CActiveAE::RunStages()
       float buftime = (float)(*it)->m_inputBuffers->m_format.m_frames / (*it)->m_inputBuffers->m_format.m_sampleRate;
       if ((*it)->m_inputBuffers->m_format.m_dataFormat == AE_FMT_RAW)
         buftime = (*it)->m_inputBuffers->m_format.m_streamInfo.GetDuration() / 1000;
```
I am still wondering if the wrong logic is just happening here, cause half of the time should be counted at this stage. That way the twice the amoung of buffers would make it without waterlevel being reached.

thexai replied:
Good, except that not explains why jumps happens also in RAW that not uses half duration. Also in v19.0 IEC that not uses 12 audio units, etc. 

Although I tested this to confirm jumps stills happening and `GetDuration` in PAPlayer still needs divided by 2 (this last really was not expecting it).

fritsch replied:
PaPlayer by itself uses a buggy concept ... the entire logic in there is flawed .. calculating one thing with samplerate and the other one with duration ... this can't work ... at least not for "half packages".

I hope to find some time over the weekend to restore one of my PCs that can do TrueHD again ... to find out, where in the stages we calculate it wrong ... PaPlayer I see a different issue. Fixing the "WATERLEVEL" might be possible - without fixing paplayer. In above equation I basically assume, that we for IEC:
- on this stage only should add up 10 ms
- that way the WATERLEVEL and time blocking will hold twice the number of packages
- Dividing by two for WATERLEVEL alone should not be needed anymore as here simply the time that was added up for comparing is now fixed.
- The real issue I think is that we only create "IEC frames" on the very last mile in the output of the sink. All the stream info and packages use / take the 10 ms version of things.

thexai replied:
> The real issue I think is that we only create "IEC frames" on the very last mile in the output of the sink. All the stream info and packages use / take the 10 ms version of things.

Then why happens also with RAW that uses 24 units / 20 ms in all stages?
Then why happens also in v18.5, v18.9, v19.0, v19.2 before MAT packer rework?

This PR fixes jumps and bad a/v sync in TrueHD in general. **(A)**

You talk about fix a different issue introduced later, only in IEC and that seems only affects PAPlayer calculations. **(B)**

Even in case to fix **B** we still needs fix **A**

fritsch replied:
I think we are more on the same page as you think.

I think you agree with me that calculating a waterlevel based on 20 ms is totally wrong, when the packages we add at this stage are only 10 ms in length. Right? This we can properly fix and make sure that we in general only calculate with 10 ms TrueHD packages. As the only place where 20 ms come into the place is when a readily packed IEC packet is given into the Sink.

The way back we need to take care, as this obviously has then a 20 ms delay which AE is not allowed to compare to its internal "10 ms" packages ...

^^ The above written is a nobrainer and I think you directly agree with me. This we have to fix to get Delay computation correct over all the stages. 

I even want to say: As long as the above describe is not fixed by us ... every "a/v" sync will just be a hack somewhere. AE needs to know what is happening, else it cannot judge. The WATERLEVEL, fluctuating or not is one symptome for that.

For PaPlayer ... yeah - as you said no matter if entirely 20 ms or 10 / 20 ms mix it has an issue in not being smooth.

Don't feel blocked by my comments. I hopefully have time over the weekend to check the entire chain.

fritsch replied:
AE measures ... times. It does not only trust package sizes it has an internal clock to know "how long stuff needed to be put on the sink" and compares a delay from that. There is definitely logical gap between what we currently "tunnel and fake" and what AE measures. It will always know that it needed 20 ms for a 10 ms packet (as it thinks it is 10 ms) at the last stage ... for this first issue AE has no "packet assumption", e.g. this 10 ms packet has a 10 ms overhead or something - it's not existing. It assumes a 1:1 relationship between packet the time it needs to write to the sink and the payload it writes.

Just to emphasize again on the first problem. PaPlayer for me is a different beast, cause as you said: It's there forever.

Emphasizing on the difference between DTS-HD-MA and TrueHD is - that we don't wait 61440 frames to create 24 packets at once. Here we work on a frame level. The old code _before_ when RAW came in, also worked on 1/24 units of TrueHD and therefore neither a 10ms implicit delay, nor a waiting for 61140 frames was needed.

So to answer your question: PaPlayer most likely was broken the very moment this 61440 workaround for the "RAW" TrueHD was introduced.

We can also solve it by dropping RAW support for TrueHD :P and work on a 1/24 unit level again :-)

thexai replied:
> I think you agree with me that calculating a waterlevel based on 20 ms is totally wrong, when the packages we add at this stage are only 10 ms in length. Right?

In AudioTrack RAW is 20 ms in all stages and waterlevel still wrong. Why?

fritsch replied:
And btw. 61440/192000 == 320 ms ... this needs to be incorporated as an offset somewhere, cause this is an additional delay that we produce by waiting to have ful 24 units.

fritsch replied:
All I want to say is: Here are multiple issues since many years and as long that is not fully understood - hacks for the waterlevel, hacks for the paplayer are needed ... and when those hacks work, users will still say: TrueHD is for me 250 ms out of sync ;-).

TrueHD fluctates, yes it does - as packages are "combined" and added only in bigger units at once - it has to fluctuate as the wallclock is ticking.
Why is the waterlevel increasing "factors" about 200 ms? Wrong calculations ... if there are 10 ms packets going in that stage - 20 frames are allowed. If those are 20 ms, 10 are allowed. If one adds 10 and the other does: 10 * GetDuration(); then the time that is compared is obviously wrong, this was what I pointed out above. This is clearly a bug - one of them.

fritsch replied:
On the one hand AE works on Duration Times (as there is nothing else you can do for RAW packages), but on the other it also works on a relation between sendframes (number of channels by 16 bit) / divided by samplerate.

PaPlayer is a nice example where one is compared / mixed with the other ... and this goes wrong for IEC (as of the new thing of 10 vs. 20) but also for RAW - cause size vs. the samplerate is not IEC padded / packed anymore and it's just wrong.

For RAW one needs to work time based. For IEC one needs to work time based AND frame based as there is a 1:1 relation ship.

Long story short: I hope to find time over the weekend to organize a TrueHD capable device to debug this.

thexai replied:
> And btw. 61440/192000 == 320 ms ... this needs to be incorporated as an offset somewhere, cause this is an additional delay that we produce by waiting to have ful 24 units.

What?   this is nonsense

61440 are bytes and TrueHD has 8 channels

61440 / 8 channels = 7860 / 2 (2 bytes or 16 bits every sample) = 3840 samples

3840 samples / 192000 samples/sec = 0.02 s = **20 ms**


**Method 2**

See _Basic Definitions_ in https://developer.dolby.com/globalassets/technology/dolby-truehd/dolbytruehdhighlevelbitstreamdescription.pdf

> The audio stream to be encoded is divided into a series of audio frames. Each audio frame contains
> the sampled data for one presentation unit. The duration of a presentation unit is **1/1200** second for
> audio sampled at multiples of 48 kHz and 1/1102.5 second for audio sampled at multiples of 44.1
> kHz

**1/1200** * 24 audio units = 0.02 = **20 ms**

fritsch replied:
Jup :-) Missed the framesize above - sorry for that :-)

fritsch replied:
Maybe you can help me: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp#L200
When I see this right - we only pack 12 Units into one 61440 frame. So that means this is the position where we actually carry only 10 ms payload but have it in a 20 ms frame?

Is this right?

fritsch replied:
Then this is the position where suddenly our "frametime" goes factor 2. Before it was 24 Units inside one 61440 byte frame, but now it's only 10. Correct? Or in other words: We generate "two calls" for OutputPacket and compute therefore 2*20ms == 40 ms while only 2 * 10 ms are consumed to do so.

fritsch replied:
We run full and should for TrueHD only return after two OutputPackage calls.

thexai replied:
> When I see this right - we only pack 12 Units into one 61440 frame. So that means this is the position where we actually carry only 10 ms payload but have it in a 20 ms frame?
> 
> Is this right?

Yes and No....   61440 only is the max capacity of buffer, audio data is padded here and can be much lower:

these  are the sizes of each audio unit:

500 | 1200 | 100 | 400 | ...... until 12 audio units

`m_dataSize`  is used only as "flag" to signal packetComplete

if m_dataSize == 0   packet is not complete

if m_dataSize == TRUEHD_BUF_SIZE  packet is "complete"    complete menas here 12 audio units

true length of audio units is calculated here:

https://github.com/xbmc/xbmc/blob/7d010efbb843fbf34018e9a5761dd74a472a338d/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp#L953-L954

As each audio unit has a header that indicates his length (coded inside audio unit as raw data). 

As you know "standard" size of each audio unit is 2560 bytes  (61440 / 24) BUT some streams can use larger audio units sporadically to make possible high bitrate peaks with condition sum of lengths of all audio units in a MAT frame ==  61440. ie is possible an audio unit of 7000 bytes followed of more small audio units (< 2560 bytes).

This true `len` is used only in MAT packer:
`m_packer->Pack(m_sinkFormat.m_streamInfo, buffer[0] + of, len);`

Padding bytes are always distributed equally between each audio unit (not at the end). This is the reason why it is a very complex system: the length of the audio units is variable and the length of the padding is also variable and dependent on all the lengths of the 24 audio units. Assuming that a very large audio unit (7000 bytes) falls right at the end of the MAT frame, the part that does not fit is saved for the next MAT frame, so there is no guarantee that data at output of  MAT packer still aligned in groups of 24 with a correspondence 1:1 as input.

But everything that I have explained here is useless to solve the problem of the jumps in PAPlayer and Bad a/v sync because as I have already mentioned about 20 times in this PR it does not depend on whether we use 12 or 24 audio units or IEC or RAW. 😊

This (jumps) already happened in **v18.5**

thexai replied:
In fact, the video that I have bothered to record that shows the jumps is using **24 audio units**.

fritsch replied:
I am not that much interested in the jumps of Paplayer. As I already mentioned what it computes is crap anyways ... by mixing bytes, samplerate and so on. You rightfully say: sucks, no matter if 24 units or 12 units.

I am concerned here:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp#L244

in combination with: 
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Audio/DVDAudioCodecPassthrough.cpp#L204

From my understanding, sorry again. This will produce "stuff" that needs 20 ms to get down the sink, the sink does not care a single bit about the content in such a frame ... it only cares how long does it need to process that data in the given configuration (8 channel 16 bit 192 khz) ... as (61440/16/192000) == 20 ms. This you compensate in this PR by adding 10 ms delay per package on top ... fine.

But the way back (!) does not fit. Somewhere someone needs to know, that "GetDuration()" of 20 ms does not mean that 20 ms of data that we currently had is now gone, but only 10 ms of them, while the rest being padding. So - the sink will always answer delay for 20 ms ... while the payload we currently have and fill is only 10 ms.

So let's go ahead and fix the paplayer scenario ... as said above hopefully weekend allows me to get that stuff reproduced ... to step through all stages.

thexai replied:
> I am not that much interested in the jumps of Paplayer.

Sorry but this is PR fixes this and nothing more. 
Maybe you shouldn't be reviewing the PR if you're not interested in the subject and don't have/use any TrueHD equipment. 

> sink will always answer delay for 20 ms ... while the payload we currently have and fill is only 10 ms.

Sink only handles MAT (IEC) and always has 24 units even when at INPUT of MAT packer has 12 units

----> 12 units + 12 units **input** ------> MAT packer --------> **output** 24 units ---------> Sink

thexai replied:
> This PR fixes jumps and bad a/v sync in TrueHD in general. (A)
> 
> You talk about fix a different issue introduced later, only in IEC and that seems only affects PAPlayer calculations. (B)
> 
> Even in case to fix B we still needs fix A
> 
> The good news is that in the case of fixing B we can drop the last commit (PAPlayer)

thexai replied:
This PR fixes **A**..... if you want can create another PR to fix **B** and drop last commit in this PR

fritsch replied:
oO -> go ahead ... if something is "fixed" - one cannot fix it again ...

fritsch replied:
![image](https://user-images.githubusercontent.com/1089309/218250947-64e0264a-64ce-442c-aef5-45962c9c2c44.png)

Finally found some time to check at the WaterLevel on a TrueHD setup. Yeah ... it is as expected most of the time exactly twice the 200 ms ... cause of AE adding 20 x 10 ms packages but GetDuration tells 20 ms ...

fritsch replied:
Good debugged further. Has absolutely nothing to do with 10 ms or 20 ms packages - it just converges to the MAX_CACHE_TIME which is 400 ms  -> all good.

Key-Question I have is the following: We basically make AE ignore WL for TrueHD. AE assumes that when the CACHE is fully filled, or at least filled more than MAX_WATER_LEVEL no correction is needed. But as TrueHD "fluctuates" quite a lot in time, we want that AE looks for TrueHD more often.

Therefore my question: Do we want to make something like this: https://github.com/fritsch/xbmc/commit/c7c6720f9381d973c951da20c611cc2941819a64

fritsch replied:
Reason: I want to avoid that short term effects will cause and AdjustError for everyone.

thexai replied:
Same here:

**12 audio units**

```
2023-02-11 11:23:38.347 T:12892    info <general>: ActiveAE::GetWaterLevel 0.280000
2023-02-11 11:23:38.355 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.355 T:12892    info <general>: ActiveAE::GetWaterLevel 0.300000
2023-02-11 11:23:38.355 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.355 T:12892    info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:23:38.355 T:12892    info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:23:38.356 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.356 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.356 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.356 T:12892    info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:23:38.356 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.356 T:12892    info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:23:38.397 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.397 T:12892    info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:23:38.399 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.399 T:12892    info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:23:38.399 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.399 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.400 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.400 T:12892    info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:23:38.400 T:12892    info <general>: Skipped 1 duplicate messages..
2023-02-11 11:23:38.400 T:12892    info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:23:38.447 T:12892    info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:23:38.447 T:12892    info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:23:38.448 T:12892    info <general>: ActiveAE::GetWaterLevel 0.280000
2023-02-11 11:23:38.454 T:12892    info <general>: Skipped 1 duplicate messages..

```

**24 audio units**

```
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.280000
2023-02-11 11:24:34.636 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.300000
2023-02-11 11:24:34.636 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.320000
2023-02-11 11:24:34.636 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.636 T:1964     info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:24:34.637 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.637 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.637 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.637 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.637 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.637 T:1964     info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:24:34.662 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.662 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.662 T:1964     info <general>: ActiveAE::GetWaterLevel 0.340000
2023-02-11 11:24:34.668 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.668 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.669 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.669 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.669 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.669 T:1964     info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:24:34.712 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.712 T:1964     info <general>: ActiveAE::GetWaterLevel 0.360000
2023-02-11 11:24:34.723 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.723 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
2023-02-11 11:24:34.723 T:1964     info <general>: Skipped 1 duplicate messages..
2023-02-11 11:24:34.723 T:1964     info <general>: ActiveAE::GetWaterLevel 0.400000
2023-02-11 11:24:34.762 T:1964     info <general>: ActiveAE::GetWaterLevel 0.380000
```

But MAIN point is that this behavior has nothing related to use 12 or 24 audio units.

fritsch replied:
Yes, see above. You are fully correct. See my suggestion. Maybe it benefits us.

thexai replied:
> Do we want to make something like this:

Sorry this not works: still jumping

fritsch replied:
Could you show me in between which values it is jumping? What is the standard deviation? Really 200 Ms at once? Here in also I don't see such heavy jumps, see also the image above

fritsch replied:
![image](https://user-images.githubusercontent.com/1089309/218255854-05c86c2e-eeee-42de-b8d1-bf40b8c9ba69.png)
From my image - I see that there is a max delta of 150 between GetWaterLevel calls. So Adjust would with my idea happen only every 100 ms jump, while the old code would jump in when waterlevel below 200 ms.

What  really want to avoid is, that the P-Regulator would bounce with the input - that way you would micro correct, which causes the opposite of what you want to achieve.

fritsch replied:
I don't want to micro discuss this issue anymore. If you would be so kind and change ignoreWL to "alwaysAdjust" with a little comment of what the issue is with TrueHD frames appearing I would already be happy :-) and we can continue to merge ... or better: test build for the Forum people.

thexai replied:
https://user-images.githubusercontent.com/58434170/218256050-25fc4604-2214-4841-8766-b1833bae40af.mp4

fritsch replied:
Okay - that's aweful ... then let's continue as it is right now ... thanks much.

thexai replied:
> or better: test build for the Forum people.

Forum feedback at this point is almost useless because now everyone there has their advancedsettings set to 90 or 100 and another negative delay in advancedsettings or GUI settings to "compensate".

At best, people are aware that they are wearing it. At worst he won't even be aware. That is, from now everyone will have a/v sync delay issues even those who have never had them :)

fritsch replied:
I think they will realize the improved A/V sync ... or realizing that now a shorter "compensate" also already helps. Or they will realize that AE now earlier adjusts and maybe earlier causes ErrorAdjust. So let's stay tuned. I posted the latest version in the forum.

fritsch commented on xbmc/cores/paplayer/PAPlayer.cpp:
```diff
@@ -860,6 +860,12 @@ inline bool PAPlayer::ProcessStream(StreamInfo *si, double &freeBufferTime)
+        if (si->m_audioFormat.m_streamInfo.m_type == CAEStreamInfo::STREAM_TYPE_TRUEHD &&
+            !m_processInfo->WantsRawPassthrough())
+        {
+          free_space /= 2;
```
This I don't fully get. Reason is: "Our frames are smaller" see below frameSent are divided by two. There I would logically assume that the "freeSpace" is twice as large.
```
free_space *= 2;
```

thexai replied:
This is not correct. See:

![vs3](https://user-images.githubusercontent.com/58434170/218254263-9bb7f424-cf27-4b89-a50a-43eca812797c.png)


But anyway i can change to form `GetDuration() / 1000 / 2;`  as seems less confusing

fritsch replied:
The confusion comes from the tunnel ... time is only half and therefore space as well ...

fritsch commented on xbmc/cores/paplayer/PAPlayer.cpp:
```diff
@@ -908,8 +914,17 @@ bool PAPlayer::QueueData(StreamInfo *si)
+      {
+        si->m_framesSent += si->m_audioFormat.m_streamInfo.GetDuration() / 1000 / 2 *
+                            si->m_audioFormat.m_streamInfo.m_sampleRate;
+      }
```
Jep fully correct.

fritsch reviewed (approved):
Thanks for the (most of the time) friendly discussing ;-)

fritsch commented:
First report is in: https://forum.kodi.tv/showthread.php?tid=371292&pid=3139700#pid3139700 (for now the only one). Let's really focus on the results - it might very well be, that through to the constant adjustment, we now micro regulate A/V by often jumping by one frame ...

thexai commented:
Obviously advancedsetting `maxpassthroughoffsyncduration` continues to do what it did before: it prevents many a/v adjustments from occurring during playback. That has not changed.

The important thing now is that even with 10 ms I can't see any flaws during playback. People who was need 90 probably still need 40 or 60 now...  this depends of how many load Shield CPU / Android OS has....

Even is possible that only by activating debug log and/or component audio CPU load increases a bit and a/v adjusts increases....

The question is: Is there a real issues in playback without debug logs enabled?  (every day use)

**10 ms debug log:**
https://paste.kodi.tv/rapeyiyala.kodi

**60 ms debug log:**
https://paste.kodi.tv/ipeharokos.kodi

thexai commented:
In truth this PR is something very simple:

It removes 10 small hacks on GetDuration() and AudioTrack and concentrates them into a single hack on water level... It's not really known yet if this is a hack or a necessary design change for TrueHD (since their implementation has unique things and may require different treatment).

What is known for sure is the other 10 changes removed were hacks.

In the PAPlayer part it is neutral: the hack that was there is still necessary (and we also know for sure that it's a hack). But it's unrelated to jumps, this only fixes that progress bar displays wrong duration (advances at 2x).

One good thing is that now all sinks on all platforms are "equal". Android is no longer the only platform that needs specific TrueHD hacks in Sink.

thexai commented:
UWP-ARM build error is unrelated (heap issues in new Windows builder)
